documentation/src/main/asciidoc/topics/upgrading.adoc

```diff
@@ -65,6 +65,19 @@ provided configuration. The possible statuses of the cache health are now HEALTH
 
 The Kryo and Protostuff marshallers have been deprecated and will be removed in Infinispan 15.0
 
+=== {brandname} now uses a version of the ProtoStream API that includes a bug
+that can affect upgrade from previous {brandname} versions. 
+
+Protobuf message types that you generate with annotation processing at compile
+time must have the same nesting in your annotated Java classes. If the nesting
+in your generated messages does not match the nesting in your classes, then
+data incompatibility issues can occur.
+
+However, the current version of the ProtoStream API does not correctly nest
+message types. For this reason, if you have Java classes that include Protobuf
+annotations, you should not nest those annotations. All Protobuf annotations in
+your Java classes should be at top-level.
+
```

Comment:
It might be easier to state that if you have previously used annotated nested classes for your types AND you have persisted these, it's necessary to refactor said classes so that they are top-level in order for the persisted data to be read.

If no stores are used, then it's not necessary todo anything :smile:

Reply:
Gotcha. I spotted your comment about binary compatibility for persisted messages. I'll revise.